tools/cdf-lineage-exporter/cdap_lineage_export/cdap/client.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import wraps
+
+from google.api_core.exceptions import NotFound
+import google.auth
+import google.auth.transport.requests
+import requests
+
+SCOPES = ["https://www.googleapis.com/auth/cloud-platform"]
+
+# Use a sentinal value so that None can be explicitly passed as credentials for
+# use in testing with local CDAP.
+# https://treyhunner.com/2019/03/unique-and-sentinel-values-in-python/
+DEFAULT_CREDENTIALS = object()
+
+
+def _empty_if_not_found(api_call):
+    """
+    Decorator to help handle scenarios where CDAP REST API returns
+    a misleading 404.
+
+    For example, If there are no datasets (or streams) the CDAP REST API
+    datasets (or streams) endpoint will return 404 instead of empty list
+    """
+
+    @wraps(api_call)
+    def handle_not_found(*args, **kwargs):
+        try:
+            return api_call(*args, **kwargs)
+        except NotFound:
+            return []
+
+    return handle_not_found
+
+
+class CDAPClient(object):
+    """
+    Client for making CDAP REST API Calls.
+
+    Args:
+        api_endpoint (str):
+            address of your CDAP instance. For Data Fusion this
+            can be retrieved with:
+
+            ..code:: shell
+
+               gcloud beta data-fusion instance describe --location=${REGION} \
+                 --format="value(apiEndpoint)" ${INSTANCE_NAME}
+
+        namespace (str):
+            Optional. Namespace to use if not specified. Defaults to
+            ``'default'``.
+
+        credentials (Optional[google.auth.credentials.Credentials]):
+            ``google.auth`` credentials object to use in requests.
+    """
+
+    def __init__(self,
+                 api_endpoint: str,
+                 credentials=DEFAULT_CREDENTIALS,
+                 namespace="default"):
+        if credentials is DEFAULT_CREDENTIALS:
+            credentials, _ = google.auth.default(scopes=SCOPES)
+            self._transport = google.auth.transport.requests.AuthorizedSession(
+                credentials)
+        else:
```

Comment:
Looks like I have a bug here for when non-`None` credentials are passed in.

```suggestion
        elif credentials is None:
```